bpf/init.sh

```diff
@@ -301,6 +303,38 @@ function bpf_load()
 	return $RETCODE
 }
 
+function bpf_load_cgroups()
+{
+	OPTS=$1
+	IN=$2
+	OUT=$3
+	PROG_TYPE=$4
+	WHERE=$5
+	SEC=$6
+	CALLS_MAP=$7
+	CGRP=$8
+
+	OPTS="${OPTS} ${OPTS_DIR} -DCALLS_MAP=${CALLS_MAP}"
+	bpf_compile $IN $OUT obj "$OPTS"
+
+	TMP=$(mktemp -d $RUNDIR/bpf.XXXXXXXXXX)
+	TMP_FILE="$TMP/x"
+	mount -t bpf bpf $TMP
+
+	cilium-map-migrate -s $OUT
+	set +e
+	tc exec bpf pin $TMP_FILE obj $OUT type $PROG_TYPE attach_type $WHERE sec $SEC
+	RETCODE=$?
+	set -e
+	cilium-map-migrate -e $OUT -r $RETCODE
+
+	if [ "$RETCODE" -eq "0" ]; then
+		bpftool cgroup attach $CGRP $WHERE pinned $TMP_FILE
```

Comment:
If this fails, then the umount two lines below will not be executed (because `-e` is set).

Reply:
Additional question: Does this replace whatever existing cgroup program is attached to the cgroup?

I'm thinking something like, if we need to recompile the base programs, we will re-run this script again and potentially attempt to `cgroup attach` here when we have previously already performed `cgroup attach` the first time the base programs were built. Could maybe test by toggling `cilium config debug=foo`.

Reply:
> Additional question: Does this replace whatever existing cgroup program is attached to the cgroup?
> 
> I'm thinking something like, if we need to recompile the base programs, we will re-run this script again and potentially attempt to `cgroup attach` here when we have previously already performed `cgroup attach` the first time the base programs were built. Could maybe test by toggling `cilium config debug=foo`.

I tested the workflow, and it's pretty much the same as with cls_bpf. I wrote in the commit description: `Attach semantics are that a previously loaded program will be overridden at the cgroups hook.`

Reply:
> If this fails, then the umount two lines below will not be executed (because `-e` is set).

Yeah, will fix that one, thanks!